components/telemetry-controller/controllers/loggingconfiguration_controller.go

```diff
@@ -0,0 +1,318 @@
+
+	telemetryv1alpha1 "github.com/kyma-project/kyma/components/telemetry-controller/api/v1alpha1"
+)
+
+const (
+	configMapFinalizer   = "FLUENT_BIT_CONFIG_MAP"
+	environmentFinalizer = "FLUENT_BIT_ENV"
+	filesFinalizer       = "FLUENT_BIT_FILES"
+)
+
+// LoggingConfigurationReconciler reconciles a LoggingConfiguration object
+type LoggingConfigurationReconciler struct {
+	client.Client
+	Scheme                  *runtime.Scheme
+	FluentBitConfigMap      types.NamespacedName
+	FluentBitDaemonSet      types.NamespacedName
+	FluentBitEnvSecret      types.NamespacedName
+	FluentBitFilesConfigMap types.NamespacedName
+}
+
+//+kubebuilder:rbac:groups=telemetry.kyma-project.io,resources=loggingconfigurations,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=telemetry.kyma-project.io,resources=loggingconfigurations/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=telemetry.kyma-project.io,resources=loggingconfigurations/finalizers,verbs=update
+//+kubebuilder:rbac:groups="",resources=configmaps,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=pods,verbs=delete;list;watch
+//+kubebuilder:rbac:groups="apps",resources=daemonsets,verbs=get;list;watch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *LoggingConfigurationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	var config telemetryv1alpha1.LoggingConfiguration
+	if err := r.Get(ctx, req.NamespacedName, &config); err != nil {
+		log.Error(err, "unable to fetch LoggingConfiguration")
+		// we'll ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	updatedCm, err := r.syncConfigMap(ctx, &config)
+	if err != nil {
+		log.Error(err, "failed to sync ConfigMap")
+		return ctrl.Result{}, err
+	}
+
+	updatedFiles, err := r.syncFiles(ctx, &config)
+	if err != nil {
+		log.Error(err, "failed to sync mounted files")
+		return ctrl.Result{}, err
+	}
+
+	updatedEnv, err := r.syncEnvironment(ctx, &config)
+	if err != nil {
+		log.Error(err, "failed to sync environment variables")
+		return ctrl.Result{}, err
+	}
+
+	if updatedCm || updatedFiles || updatedEnv {
+		if err := r.Update(ctx, &config); err != nil {
+			log.Error(err, "cannot update LoggingConfiguration")
+			return ctrl.Result{}, err
+		}
+
+		if err := r.deleteFluentBitPods(ctx, log); err != nil {
+			log.Error(err, "cannot delete fluent bit pods")
+			return ctrl.Result{}, err
+		}
```

Comment:
What if instead we trigger a rolling upgrade of the Fluent Bit `DaemonSet` (by applying a dummy label or annotation. or something) to have the rollout performed in a controllable fashion?

Reply:
I would like to avoid adding an annotation as long as the DaemonSet is managed externally, which would potentially cause unnecessary restarts. Also the `kubectl rollout restart` could solve this without any modifications. However, this is implemented by some patch operation and since the test suite does not have a kube-controller-manager it would be hard to test.